src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -0,0 +1,429 @@
+    } finally {
+      flushRequested = false;
+    }
+  }
+
+  private final class BulkTask implements Callable<BulkResponse> {
+
+    final long batchId = createdBatches.incrementAndGet();
+
+    final List<R> batch;
+
+    BulkTask(List<R> batch) {
+      this.batch = batch;
+    }
+
+    @Override
+    public BulkResponse call() throws Exception {
+      final BulkResponse rsp;
+      try {
+        rsp = execute();
+      } catch (Exception e) {
+        failAndStop(e);
+        throw e;
+      }
+      log.debug("Successfully executed batch {} of {} records", batchId, batch.size());
+      onBatchCompletion(batch.size());
+      return rsp;
+    }
+
+    private BulkResponse execute() throws Exception {
+      final B bulkReq;
+      try {
+        bulkReq = bulkClient.bulkRequest(batch);
+      } catch (Exception e) {
+        log.error("Failed to create bulk request from batch {} of {} records", batchId, batch.size(), e);
+        throw e;
+      }
+      for (int remainingRetries = maxRetries; true; remainingRetries--) {
+        boolean retriable = true;
+        try {
+          log.trace("Executing batch {} of {} records", batchId, batch.size());
+          final BulkResponse bulkRsp = bulkClient.execute(bulkReq);
+          if (bulkRsp.isSucceeded()) {
+            return bulkRsp;
+          }
+          retriable = bulkRsp.isRetriable();
+          throw new ConnectException("Bulk request failed: " + bulkRsp.getErrorInfo());
+        } catch (Exception e) {
+          if (retriable && remainingRetries > 0) {
+            log.warn("Failed to execute batch {} of {} records, retrying after {} ms", batchId, batch.size(), retryBackoffMs, e);
+            time.sleep(retryBackoffMs);
+          } else {
+            log.error("Failed to execute batch {} of {} records", batchId, batch.size(), e);
+            throw e;
+          }
+        }
+      }
+    }
+
+  }
+
+  private synchronized void onBatchCompletion(int batchSize) {
+    successfulBatches.incrementAndGet();
+    successfulRecords.addAndGet(batchSize);
+    inFlightRecords -= batchSize;
+    assert inFlightRecords >= 0;
+    notifyAll();
+  }
+
+  private void failAndStop(Throwable t) {
+    error.compareAndSet(null, toConnectException(t));
+    stop();
+  }
+
+  /**
+   * @return count of currently buffered records
+   */
+  public synchronized int unsentRecords() {
+    return unsentRecords.size();
+  }
+
+  /**
+   * @return count of records currently in flight
+   */
+  public synchronized int inFlightRecords() {
+    return inFlightRecords;
+  }
+
+  /**
+   * @return sum of unsent and in-flight record counts
+   */
+  public synchronized int bufferedRecords() {
+    return unsentRecords.size() + inFlightRecords;
+  }
+
+  /**
+   * @return count of batches that have been created
+   */
+  public long createdBatches() {
+    return createdBatches.get();
+  }
+
+  /**
+   * @return count of batches successfully executed
+   */
+  public long successfulBatches() {
```

Comment:
Without these unsynchronized accessors, it doesn't look like we need `successfulBatches` and `successfulRecords` to be `AtomicLong`s since they are only touched in the (`synchronized`) `onBatchCompletion`.